src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs

```diff
@@ -308,6 +310,27 @@ private static ArraySegment<byte> ReadToEnd(Stream stream)
 
                 int lastRead;
 
+                // Read up to 3 bytes to see if it's the UTF-8 BOM
+                do
+                {
+                    // No need for checking for growth, the minimal rent sizes both guarantee it'll fit.
+                    Debug.Assert(rented.Length >= utf8Bom.Length);
+
+                    lastRead = stream.Read(
+                        rented,
+                        written,
+                        utf8Bom.Length - written);
```

Comment:
Why not read up to `rented.Length - written` to avoid an extra "small" read?

And then, we can just check if `rented.StartsWith(utf8Bom)`

Reply:
The original version that I wrote just checked at the end if there was a BOM and changed the ReadOnlyMemory to use a non-zero start index; the problem is that the "I'm renting data on behalf of a document, clear it" doesn't know what the start index was.

The solutions that I envisioned:

* Make JsonDocument bigger by storing the start position of a temporary rent
* Make temporary rents clear the whole array before returning it, instead of just the written to portion
* Use Buffer.BlockCopy to move the data backwards back to position 0.
* This one

"small read" seemed the lowest impact.